### ISSUE TITLE
browser integration regression with 2.6.0 (keepassxc-proxy)

### ISSUE THREAD

**Issue by leo-b** [flathub/org.keepassxc.KeePassXC#67]
The flathub version of keepassxc 2.6.0 doesn't create a socket file for communication with the browser extension anymore.

Up to 2.5.3, a socket file ``/run/user/<userid>/app/org.keepassxc.KeePassXC/kpxc_server`` was created if _Enable browser integration_ was selected in keepassxc settings.
``keepasssx-wrapper`` used this socket in order to detect wether it should run ``keepassxc`` or ``keepassxc-proxy``.

In 2.6.0 ``keepasssx-wrapper`` looks for a different file name: ``/run/user/<userid>/app/org.keepassxc.KeePassXC/org.keepassxc.KeePassXC.BrowserServer`` but no socket is created at all.
As a result ``keepassxc`` gets run instead of ``keepassxc-proxy`` which exits immediately and results in a loop.

If I remove the check for the socket file in keepassxc-wrapper, ``keepassxc-proxy`` gets launched but since the socket is missing, no communication is possible.

**Comment by AsavarTzeth** [comment]
Strange and on my system there is a socket but it uses the old name `org.keepassxc.KeePassXC/kpxc_server` and it still works. Different from what is expected and wrong multiple ways. I wonder, what I may have messed up here?

Looking at https://github.com/keepassxreboot/keepassxc/commit/2237cf0188af38898e1492f5d7ce2c0f087c8f35 and https://github.com/keepassxreboot/keepassxc/blob/2.6.0/src/browser/BrowserShared.cpp#L30 indicates that new name should be used.

The funny thing is I don't even touch this variable in my patches.

I tried uninstalling and installing the app and now everything works as expected. Perhaps for whatever reason there is an issue that prevented a clean update?

**Comment by leo-b** [comment]
I have some new insights (tested with 2.6.1) and a workaround for my problem.

I noticed that in my setup the unix socket is created as
``/run/user/5020/app/org.keepassxc.KeePassXC.BrowserServer``
as opposed to
``/run/user/5020/app/org.keepassxc.KeePassXC/org.keepassxc.KeePassXC.BrowserServer
``
which I'd expect looking at the source.

That's the reason why _keepassxc-wrapper_ cannot find the socket and thus decides to run _keepassxc_ instead of _keepassxc-proxy_:
```
# Browser integration is enabled if unix socket exists
if [[ -S "${XDG_RUNTIME_DIR}/app/${appId}/${appId}.BrowserServer" ]]; then
  if [[ "$1" =~ "${arg1}" ]] || [[ "$2" =~ "${arg2}" ]]; then
    exec keepassxc-proxy "$@"
  fi
fi

# If no arguments are matched or browser integration is off execute keepassxc
exec keepassxc "$@"
```

As a second issue, the socket cannot be created as ``{XDG_RUNTIME_DIR}/app/${appId}.BrowserServer`` because according to
https://docs.flatpak.org/en/latest/sandbox-permissions.html ``/run`` is
blacklisted with the exception of ``$XDG_RUNTIME_DIR/app/$FLATPAK_ID``:

> No access to any host files except the runtime, the app,~/.var/app/$FLATPAK_ID, and $XDG_RUNTIME_DIR/app/$FLATPAK_ID. Only the latter two being writable.

Thus I added a filesystem override to this path as a workaround:
```
flatpak override --system --filesystem=xdg-run/app org.keepassxc.KeePassXC
```
... which I guess isn't a very good idea to leave as a permanent solution.

In addition I patched ``keepassxc-wrapper`` to look for the other socket path:
```
--- keepassxc-wrapper.orig      2020-10-02 11:01:57.416251077 +0200
+++ keepassxc-wrapper   2020-10-02 11:02:05.988546755 +0200
@@ -31,7 +31,7 @@
 readonly arg2='keepassxc-browser@keepassxc.org'
 
 # Browser integration is enabled if unix socket exists
-if [[ -S "${XDG_RUNTIME_DIR}/app/${appId}/${appId}.BrowserServer" ]]; then
+if [[ -S "${XDG_RUNTIME_DIR}/app/${appId}.BrowserServer" ]]; then
   if [[ "$1" =~ "${arg1}" ]] || [[ "$2" =~ "${arg2}" ]]; then
     exec keepassxc-proxy "$@"
   fi
```

Using this workaround, the browser extension is able to connect to keepassxc again.

However this is not a solution to the problem and I still have no clue why keepassxc uses the wrong socket path.
Do you have any hint?

**Comment by AsavarTzeth** [comment]
I can see you have done your research.

While this remains unreproducible on my own systems I do have one idea. What do you get if you run:

    flatpak run --command='env' org.keepassxc.KeePassXC | grep FLATPAK_ID

The only thing I could think of is that variable being empty or unset for some reason. I really expected that environment variable to be reliable within the sandbox. If it isn't then I could just go back to hard coding the id and now with a good excuse to do so.

For reference, if `FLATPAK_ID` is broken (for whatever reason) this would be the problematic lines:
https://github.com/flathub/org.keepassxc.KeePassXC/blob/945a4bc3ae36f60d9d96ce5b0ff833df58e29544/patch/keepassxc/0003-Flatpak-Support-KeePassXC-Browser-integration.patch#L70-L80

**Comment by leo-b** [comment]
It looks like you are right and ``FLATPAK_ID`` really isn't set on my system:

```
$ flatpak run --command='env' org.keepassxc.KeePassXC | grep FLATPAK
FLATPAK_SANDBOX_DIR=/afs/wu-wien.ac.at/home/edvz/bergolth/.var/app/org.keepassxc.KeePassXC/sandbox
```

The running keepassxc process also doesn't have this variable set:
```
# ps 15979
  PID TTY      STAT   TIME COMMAND
15979 pts/2    SLl    0:03 keepassxc
# cat /proc/15979/environ |tr '\0' '\n' |sort | grep FLATPAK
FLATPAK_SANDBOX_DIR=/afs/wu-wien.ac.at/home/edvz/bergolth/.var/app/org.keepassxc.KeePassXC/sandbox
```

Maybe ``FLATPAK_ID`` was only introduced in recent flatpak versions? I am running Centos 7 which includes flatpak 1.0.9.
```
# rpm -q flatpak     
flatpak-1.0.9-9.el7_7.x86_64
```

**Comment by AsavarTzeth** [comment]
I believe you are correct. Looking closer at the origin of `FLATPAK_ID`, I found it was initially introduced in https://github.com/flatpak/flatpak/commit/bbdf2d11c8b6a3aecc57cb7760daf2fedc28d919 and later expanded to **flatpak-run(1)** in https://github.com/flatpak/flatpak/commit/f9a73904bd10643a3ee7b2cd4141138c11d9e2d2. It was first included in 1.1.2 (pre-release for 1.2.0).

In 1c8bbc6 I have reverted to hard coding the id whenever possible. Changing from this was in retrospect a bad idea and a clear mistake on my part, regardless if it may have worked for most users or not.

You were a great help in locating the source of this.
